# 1. Первичный анализ данных

## 🎯 Цель
Провести первичный осмотр данных: понять структуру, типы признаков, наличие пропусков и дубликатов

## 🔧 Методы
- `df.shape` - размерность данных
- `df.info()` - структура и типы данных
- `df.describe()` - статистики числовых признаков
- `df.isna().sum()` - проверка пропусков
- `df.duplicated().sum()` - проверка дубликатов
- `df.nunique()` - количество уникальных элементов

## 🗂️ Описание признаков

| Признак | Тип | Описание |
|---------|-----|----------|
| `RowNumber` | int64 | Номер строки |
| `CustomerId` | int64 | Уникальный ID клиента |
| `Surname` | object | Фамилия клиента |
| `CreditScore` | int64 | Кредитный рейтинг |
| `Geography` | object | Страна (Germany/France/Spain) |
| `Gender` | object | Пол (Male/Female) |
| `Age` | int64 | Возраст |
| `Tenure` | int64 | Количество лет, в течение которых клиент работает в банке |
| `Balance` | float64 | Баланс на счете |
| `NumOfProducts` | int64 | Количество продуктов банка |
| `HasCrCard` | int64 | Наличие кредитной карты (0/1) |
| `IsActiveMember` | int64 | Активный член (0/1) |
| `EstimatedSalary` | float64 | Примерная зарплата |
| `Exited` | int64 | **Целевая**: Ушел из банка (0/1) |

## 👀 Первый взгляд на данные

Ознакомление с реальными значениями признаков, проверка корректности данных.

In [9]:
import pandas as pd

df = pd.read_csv('../data/Churn_Modelling.csv')
df.head(20)

RowNumber  CustomerId    Surname  CreditScore Geography  Gender    Age  \
0           1    15634602   Hargrave          619    France  Female  42.00   
1           2    15647311       Hill          608     Spain  Female  41.00   
2           3    15619304       Onio          502    France  Female  42.00   
3           4    15701354       Boni          699    France  Female  39.00   
4           5    15737888   Mitchell          850     Spain  Female  43.00   
5           6    15574012        Chu          645     Spain    Male  44.00   
6           7    15592531   Bartlett          822       NaN    Male  50.00   
7           8    15656148     Obinna          376   Germany  Female  29.00   
8           9    15792365         He          501    France    Male  44.00   
9          10    15592389         H?          684    France    Male    NaN   
10         11    15767821     Bearce          528    France    Male  31.00   
11         12    15737173    Andrews          497     Spain    Male  24.00   
12         13    15632264        Kay          476    France  Female  34.00   
13         14    15691483       Chin          549    France  Female  25.00   
14         15    15600882      Scott          635     Spain  Female  35.00   
15         16    15643966    Goforth          616   Germany    Male  45.00   
16         17    15737452      Romeo          653   Germany    Male  58.00   
17         18    15788218  Henderson          549     Spain  Female  24.00   
18         19    15661507    Muldrow          587     Spain    Male  45.25   
19         20    15568982        Hao          726    France  Female  24.00   

    Tenure    Balance  NumOfProducts  HasCrCard  IsActiveMember  \
0        2       0.00              1        1.0             1.0   
1        1   83807.86              1        0.0             1.0   
2        8  159660.80              3        1.0             0.0   
3        1       0.00              2        0.0             0.0   
4        2  125510.82              1        NaN             1.0   
5        8  113755.78              2        1.0             0.0   
6        7       0.00              2        1.0             1.0   
7        4  115046.74              4        1.0             0.0   
8        4  142051.07              2        0.0             NaN   
9        2  134603.88              1        1.0             1.0   
10       6  102016.72              2        0.0             0.0   
11       3       0.00              2        1.0             0.0   
12      10       0.00              2        1.0             0.0   
13       5       0.00              2        0.0             0.0   
14       7       0.00              2        1.0             1.0   
15       3  143129.41              2        0.0             1.0   
16       1  132602.88              1        1.0             0.0   
17       9       0.00              2        1.0             1.0   
18       6       0.00              1        0.0             0.0   
19       6       0.00              2        1.0             1.0   

    EstimatedSalary  Exited  
0         101348.88       1  
1         112542.58       0  
2         113931.57       1  
3          93826.63       0  
4          79084.10       0  
5         149756.71       1  
6          10062.80       0  
7         119346.88       1  
8          74940.50       0  
9          71725.73       0  
10         80181.12       0  
11         76390.01       0  
12         26260.98       0  
13        190857.79       0  
14         65951.65       0  
15         64327.26       0  
16          5097.67       1  
17         14406.41       0  
18        158684.81       0  
19         54724.03       0

## 📏 Определяем размер датасета

In [16]:
df.shape

(10002, 14)

## ℹ️Выведем основную информацию по датасету

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10002 entries, 0 to 10001
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   RowNumber        10002 non-null  int64  
 1   CustomerId       10002 non-null  int64  
 2   Surname          10002 non-null  object 
 3   CreditScore      10002 non-null  int64  
 4   Geography        10001 non-null  object 
 5   Gender           10002 non-null  object 
 6   Age              10001 non-null  float64
 7   Tenure           10002 non-null  int64  
 8   Balance          10002 non-null  float64
 9   NumOfProducts    10002 non-null  int64  
 10  HasCrCard        10001 non-null  float64
 11  IsActiveMember   10001 non-null  float64
 12  EstimatedSalary  10002 non-null  float64
 13  Exited           10002 non-null  int64  
dtypes: float64(5), int64(6), object(3)
memory usage: 1.1+ MB


## 📊 Результаты info()
- **Записей:** 10,002
- **Признаков:** 14
- **Пропуски:** ✅ присутствуют по одному пропуску в столбцах "Geography", "Age", "HasCrCard", "IsActiveMember"
- **Типы данных:**
  - `int64`: 6 признаков (включая целевую `Exited`)
  - `object`: 3 признака (`Geography`, `Gender`, `Surname` - категориальные)
  - `float64`: 5 признак (`Age`, `Balance`, `HasCrCard`, `IsActiveMember`, `EstimatedSalary`)

## 📈 Статистики числовых признаков

Анализируем распределение числовых переменных: средние, квартили, разброс значений.

In [7]:
df.describe()

RowNumber    CustomerId   CreditScore           Age        Tenure  \
count  10002.000000  1.000200e+04  10002.000000  10001.000000  10002.000000   
mean    5001.499600  1.569093e+07    650.555089     38.922311      5.012498   
std     2887.472338  7.193177e+04     96.661615     10.487200      2.891973   
min        1.000000  1.556570e+07    350.000000     18.000000      0.000000   
25%     2501.250000  1.562852e+07    584.000000     32.000000      3.000000   
50%     5001.500000  1.569073e+07    652.000000     37.000000      5.000000   
75%     7501.750000  1.575323e+07    718.000000     44.000000      7.000000   
max    10000.000000  1.581569e+07    850.000000     92.000000     10.000000   

             Balance  NumOfProducts     HasCrCard  IsActiveMember  \
count   10002.000000   10002.000000  10001.000000    10001.000000   
mean    76491.112875       1.530194      0.705529        0.514949   
std     62393.474144       0.581639      0.455827        0.499801   
min         0.000000       1.000000      0.000000        0.000000   
25%         0.000000       1.000000      0.000000        0.000000   
50%     97198.540000       1.000000      1.000000        1.000000   
75%    127647.840000       2.000000      1.000000        1.000000   
max    250898.090000       4.000000      1.000000        1.000000   

       EstimatedSalary        Exited  
count     10002.000000  10002.000000  
mean     100083.331145      0.203759  
std       57508.117802      0.402812  
min          11.580000      0.000000  
25%       50983.750000      0.000000  
50%      100185.240000      0.000000  
75%      149383.652500      0.000000  
max      199992.480000      1.000000

## 📈 Основные выводы из describe()

**По числовым признакам:**
- `CreditScore`: распределение 300-850 (типично для кредитных скорингов)
- `Age`: от 18 до 92 лет, потенциальные выбросы в старшем возрасте
- `Balance`: от 0 до 250K, много нулевых значений (требует проверки)
- `EstimatedSalary`: равномерное распределение

**Требуют внимания в EDA:**
- Проверить распределение Age на выбросы
- Исследовать нулевые Balance
- Проанализировать экстремальные значения CreditScore

## ⚠️Проверка на пропуски в данных

Проверим сколько пропусков имеется в нашем датасете

In [10]:
df.isna().sum()

RowNumber          0
CustomerId         0
Surname            0
CreditScore        0
Geography          1
Gender             0
Age                1
Tenure             0
Balance            0
NumOfProducts      0
HasCrCard          1
IsActiveMember     1
EstimatedSalary    0
Exited             0
dtype: int64

## 🔍 Проверка на дубликаты

Ищем полностью повторяющиеся строки, которые могут искажать анализ.

In [13]:
df.duplicated().sum()

2

## 🔢Проверка уникальных значений

Смотрим, сколько у нас уникальных значений

In [14]:
df.nunique()

RowNumber          10000
CustomerId         10000
Surname             2932
CreditScore          460
Geography              3
Gender                 2
Age                   73
Tenure                11
Balance             6382
NumOfProducts          4
HasCrCard              2
IsActiveMember         2
EstimatedSalary     9999
Exited                 2
dtype: int64

## 🎯 Итоги первичного анализа

**✅ Качество данных:**
- 10,002 записи, 14 признаков
- Минимальные пропуски (по 1 в 4 столбцах)
- Дубликаты отсутствуют

**🔍 Находки для EDA:**
- Сбалансированность целевой переменной `Exited`
- Потенциальные выбросы в `Age`
- Много нулевых значений в `Balance`

**📋 План следующих действий:**
1. Детальный анализ распределений (Этап 2 - EDA)
2. Исследование корреляций
3. Анализ влияния категориальных признаков